### DISCUSSION TITLE
A dependency expanded by @DependencyClient macro can not be constructed outside itself

### DISCUSSION THREAD

**Discussion by iharandreyev** [pointfreeco/swift-dependencies#297] [General]
### Description

This is my first time trying @DependencyClient macro.

If the macro is removed, swift generates a constructor with parameters, and if `ValueGenerator<Output>()` is not used (the generated constructor with no parameters), things compile.
If macro is applied, upon expanding it, there seems to be two constructors as claimed by the documentation, but they are not available outside of `ValueGenerator` scope for some reason.

BTW, the same happens if 'FooReducer.State' is not declared explicitly.
An empty 'State.init()' also can't be invoked, the compiler expects 'State(from:)' decodable init.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-dependencies/issues) or [discussion](https://github.com/pointfreeco/swift-dependencies/discussions).

### Expected behavior

The sample code compiles successfully

### Actual behavior

The sample code does not compile with `'ValueGenerator<Int>' cannot be constructed because it has no accessible initializers error`.

### Steps to reproduce

```
@DependencyClient
struct ValueGenerator<Output> {
  var generateValue: @Sendable () async throws -> Output
}

extension ValueGenerator: TestDependencyKey {
  static var testValue: ValueGenerator<Output> {
    return ValueGenerator<Output>()
  }
}

@Reducer
struct FooReducer {
  @ObservableState
  struct State: Equatable {
      init() { }
  }
  
  @Dependency(ValueGenerator<Int>.self)
  var generator
}

struct FooView: View {
  let store: StoreOf<FooReducer>
  
  var body: some View {
    Text("Hello World!")
  }
}

#Preview {
  FooView(
    store: Store(
      initialState: FooReducer.State(),
      reducer: FooReducer.init,
      withDependencies: { values in
        // The line below fails with "'ValueGenerator<Int>' cannot be constructed because it has no accessible initializers" error
        let generator = ValueGenerator<Int>(
          generateValue: {
            return 1
          }
        )
        values[ValueGenerator<Int>.self] = generator
        )
      }
    )
  )
}
```

### Dependencies version information

1.4.1

### Destination operating system

14.0

### Xcode version information

16.1 (16B40)

### Swift Compiler version information

```shell
% swift --version
swift-driver version: 1.115 Apple Swift version 6.0.2 (swiftlang-6.0.2.1.2 clang-1600.0.26.4)
Target: arm64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @iharandreyev, the code you have provided compiles just fine for me. Is all of this code in the same file? Or are you try to construct the `ValueGenerator` from a different module? If so, then you need to make `ValueGenerator` and `generateValue` public.

Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!

**Reply by iharandreyev** [reply]
@mbrandonw The issue happens in the same file.
Maybe there are some sort of permissions the macro needs, that are not obvious?

**Reply by iharandreyev** [reply]
@mbrandonw 
Alright, I have found an answer.
Seems like it has to do with deployment target being `iOS 14.0`, and `#Preview` macro only being available since` iOS 17`. 
If using old `PreviewProvider` construct, things work as expected.
There's a discussion on [stack overflow](https://stackoverflow.com/questions/77850121/availability-checks-break-preview-macro-with-nested-type-declaration) regarding the issue.

**Reply by mbrandonw** [reply]
Oh I see. I didn’t bring over the preview code when I tested. This is happening because macros can’t see code that other macros generate. So really no macro generated code can be accessed inside `#Preview`, ever. It’s not an iOS version thing. It’s just how macros work unfortunately.
